Frameworks/CoreGraphics/CGContextCairo.mm

```diff
@@ -1633,6 +1634,358 @@
+
+    // Calculates points on stroked path from original points
+    // Reverse flag should be true if we're traversing the path backwards
+    auto strokedPoint = [](CGPoint p, CGFloat angle, CGFloat distance, bool reverse) -> CGPoint {
+        if (reverse) {
+            angle += M_PI;
+        }
+        CGFloat x = p.x - (sin(angle) * distance);
+        CGFloat y = p.y + (cos(angle) * distance);
+
+        return CGPointMake(x, y);
+    };
+
+    // Adds a line cap
+    auto addLineCap = [halfStrokeWidth, &strokedPoint](cairo_t* context, CGFloat angle, CGPoint p1, CGPoint p2) -> void {
+        cairo_line_cap_t lineCap = cairo_get_line_cap(context);
+        switch (lineCap) {
+            case CAIRO_LINE_CAP_ROUND: {
+                // Add arc to close sub path
+                CGPoint arcMidPoint = CGPointMake((p1.x + p2.x) / 2.0, (p1.y + p2.y) / 2.0);
+                cairo_arc_negative(context, arcMidPoint.x, arcMidPoint.y, halfStrokeWidth, angle + M_PI_2, angle - M_PI_2);
+                break;
+            }
+
+            case CAIRO_LINE_CAP_BUTT: {
+                // Draw a line segment to close sub path
+                cairo_line_to(context, p2.x, p2.y);
+                break;
+            }
+
+            case CAIRO_LINE_CAP_SQUARE:
+                // Add line segments for square at end of sub path
+                CGPoint p1a = strokedPoint(p1, angle - M_PI_2, halfStrokeWidth, false);
+                CGPoint p2a = strokedPoint(p2, angle - M_PI_2, halfStrokeWidth, false);
+                cairo_line_to(context, p1a.x, p1a.y);
+                cairo_line_to(context, p2a.x, p2a.y);
+                cairo_line_to(context, p2.x, p2.y);
+                break;
+        }
+    };
+
+    auto addLineJoin = [halfStrokeWidth,
+                        &strokedPoint,
+                        this](cairo_t* context, CGFloat angle1, CGPoint p1, CGFloat angle2, CGPoint p2, bool reverse) -> void {
+
+        double miterLimit = cairo_get_miter_limit(context);
+        cairo_line_join_t lineJoin = cairo_get_line_join(context);
+
+        // Check to see if join should be overridden to use bevel
+        CGFloat delta = angle2 - angle1;
+        delta = fmod(delta, 2.0f * M_PI);
+        if (delta < 0.0f) {
+            delta += 2.0f * M_PI;
+        }
+        // Force acute angles to use bevel
+        // Top range is temporary fix to avoid joins
+        // on intermediate curve lines
+        if (delta < M_PI || delta > (1.98f * M_PI)) {
+            lineJoin = CAIRO_LINE_JOIN_BEVEL;
+        }
+
+        if (lineJoin == CAIRO_LINE_JOIN_MITER) {
+            CGFloat m1 = tan(angle1);
+            CGFloat b1 = p1.y - (m1 * p1.x);
+
+            CGFloat m2 = tan(angle2);
+            CGFloat b2 = p2.y - (m2 * p2.x);
+
+            CGFloat x = (b2 - b1) / (m1 - m2);
+            CGFloat y = 0;
+            // If m1 is very large it's probably a vertical line
+            if (m1 > 1000000) {
```

Comment:
Date: Tue, 26 Jul 2016 06:50:54 -0700
From: notifications@github.com
To: WinObjC@noreply.github.com
Subject: Re: [Microsoft/WinObjC] Basic ReplacePathWithStrokedPath Implementation (#625)

In Frameworks/CoreGraphics/CGContextCairo.mm:

> +        // on intermediate curve lines
> +        if (delta < M_PI || delta > (1.98f * M_PI)) {
> +            lineJoin = CAIRO_LINE_JOIN_BEVEL;
> +        }
> +
> +        if (lineJoin == CAIRO_LINE_JOIN_MITER) {
> +            CGFloat m1 = tan(angle1);
> +            CGFloat b1 = p1.y - (m1 * p1.x);
> +
> +            CGFloat m2 = tan(angle2);
> +            CGFloat b2 = p2.y - (m2 * p2.x);
> +
> +            CGFloat x = (b2 - b1) / (m1 - m2);
> +            CGFloat y = 0;
> +            // If m1 is very large it's probably a vertical line
> +            if (m1 > 1000000) {

The number is arbitrary, m1 > 1000000 can be replaced with angle1 == M_PI / 2 || angle1 == M_PI * 3 / 2 (with float approximations).


—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.BV JNMVFEBJLFF,MFFCMFFEHEHNERHRNFHHFRTNN VFHNFNFBHGHNHGJLFGTUI;TGTKIVGJKGVFGFNHFBGFFHFHFVFB FBFHBFVFBFBFBHNVBVG ED,DHNEDBEBEHNN NMMMNVRRMNJRNNRR RNRRMM  GMFM F F VGNVGN BEEEEBENEBEHJNGVGV NVG     VNLGNFFJNFNNFVVFNNVGGFNVGFN VFG NVFG N
VFG MM G G  B B B M